File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Solutions/RunAnywhere+Solutions.swift
//
//  RunAnywhere+Solutions.swift
//  RunAnywhere SDK
//
//  Public API for L5 solutions runtime (T4.7 / T4.8). A "solution" is a
//  prepackaged pipeline config — either a typed `RASolutionConfig` proto
//  or YAML sugar — that the C++ core compiles into a GraphScheduler DAG
//  and executes through the `rac_solution_*` C ABI.
//
//  Surface mirrors the rest of the SDK's capability-shape: callers reach
//  the API through `RunAnywhere.solutions.run(config:)`. The proto-byte
//  variant is the canonical entry point; the typed-proto helper is a
//  thin convenience that serialises and forwards.
//

import CRACommons
import Foundation
import os
import SwiftProtobuf

// MARK: - SolutionHandle

/// Opaque, ARC-safe wrapper around a `rac_solution_handle_t`.
///
/// Owns the underlying C handle and guarantees `rac_solution_destroy`
/// runs exactly once (either explicitly via `destroy()` or on the final
/// release through `deinit`). All lifecycle verbs are forwarded one-to-one
/// to the C ABI.
public final class SolutionHandle: @unchecked Sendable {

    // Per AGENTS.md: NSLock is forbidden — use `OSAllocatedUnfairLock`.
    private let handle: OSAllocatedUnfairLock<SolutionHandleState>

    // swiftlint:disable:next strict_fileprivate
    fileprivate init(handle: rac_solution_handle_t) {
        self.handle = OSAllocatedUnfairLock(
            initialState: SolutionHandleState(handle: handle)
        )
    }

    deinit {
        handle.withLock { state in
            if let current = state.handle {
                rac_solution_destroy(current)
                state.handle = nil
            }
        }
    }

    /// Start the underlying scheduler. Non-blocking.
    public func start() throws {
        try withHandle { rac_solution_start($0) }
    }

    /// Request a graceful shutdown. Non-blocking.
    public func stop() throws {
        try withHandle { rac_solution_stop($0) }
    }

    /// Force-cancel the graph. Returns once worker threads observe cancellation.
    public func cancel() throws {
        try withHandle { rac_solution_cancel($0) }
    }

    /// Feed a single UTF-8 item into the root input edge.
    public func feed(_ item: String) throws {
        try withHandle { handle in
            item.withCString { rac_solution_feed(handle, $0) }
        }
    }

    /// Signal end-of-stream on the root input edge.
    public func closeInput() throws {
        try withHandle { rac_solution_close_input($0) }
    }

    /// Cancel, join, and destroy the solution. Idempotent.
    public func destroy() {
        handle.withLock { state in
            if let current = state.handle {
                rac_solution_destroy(current)
                state.handle = nil
            }
        }
    }

    /// Whether the underlying C handle is still live (CANONICAL_API §11).
    ///
    /// Returns `true` as long as `destroy()` has not been called (or `deinit`
    /// triggered). Once destroyed, further calls to the lifecycle verbs throw
    /// `SDKException(.invalidState)`.
    public var isAlive: Bool {
        handle.withLock { $0.handle != nil }
    }

    /// Run `body` against the native handle while holding the slot lock so
    /// that a concurrent `destroy()` (or `deinit`) cannot free the handle
    /// mid-call. `rac_solution_destroy` cancels and joins before deleting
    /// the runner, so serialising the lifecycle verbs is the simplest way
    /// to honour the Sendable contract without resurrecting a freed pointer
    /// (see swift-public-features-002).
    private func withHandle(
        _ body: @Sendable (rac_solution_handle_t) -> rac_result_t
    ) throws {
        let result: rac_result_t = try handle.withLock { state in
            guard let current = state.handle else {
                throw SDKException(
                    code: .invalidState,
                    message: "Solution handle has already been destroyed",
                    category: .internal
                )
            }
            return body(current)
        }
        guard result == RAC_SUCCESS else {
            throw SDKException(
                code: .processingFailed,
                message: SolutionHandle.errorMessage(
                    op: "Solution lifecycle call",
                    rc: result
                ),
                category: .internal
            )
        }
    }

    // Compose a human-readable failure message from a `rac_result_t`.
    //
    // Pairs `rac_error_message(rc)` (canonical static description per
    // code) with `rac_error_get_details()` (thread-local detail set by
    // the failing C call, if any) so callers see the underlying cause
    // instead of just a numeric code.
    // swiftlint:disable:next strict_fileprivate
    fileprivate static func errorMessage(op: String, rc: rac_result_t) -> String {
        var message = "\(op) failed"
        let description = String(cString: rac_error_message(rc))
        if !description.isEmpty {
            message += ": \(description)"
        }
        if let detailPtr = rac_error_get_details() {
            let detail = String(cString: detailPtr)
            if !detail.isEmpty {
                message += " (\(detail))"
            }
        }
        message += " [rc=\(rc)]"
        return message
    }
}

/// The native solution pointer is owned by `SolutionHandle` and is only read,
/// invoked, or cleared while the enclosing unfair lock is held.
private struct SolutionHandleState: @unchecked Sendable {
    var handle: rac_solution_handle_t?
}

// MARK: - Solutions Capability

public extension RunAnywhere {

    /// Capability accessor for solution-runtime operations.
    ///
    /// Mirrors the Kotlin/Flutter/RN/Web shape — `RunAnywhere.solutions.run(config:)`.
    static var solutions: Solutions { Solutions() }

    /// Stateless namespace for solution-runtime APIs. Backed by the C ABI;
    /// keeps no mutable state of its own — every handle returned by `run`
    /// owns its own native solution.
    struct Solutions: Sendable {

        // swiftlint:disable:next strict_fileprivate
        fileprivate init() {}

        /// Construct and return a started solution from a serialised
        /// `runanywhere.v1.SolutionConfig` (or `PipelineSpec`) message.
        ///
        /// The handle is returned in the **created** state — call
        /// `start()` to launch worker threads. Callers that only want the
        /// "create + start" combo should chain `try handle.start()`.
        ///
        /// - Parameter configBytes: Serialized `SolutionConfig` proto.
        /// - Returns: Lifecycle handle owning the underlying solution.
        public func run(configBytes: Data) async throws -> SolutionHandle {
            try await ensureReady()

            var raw: rac_solution_handle_t?
            let result = configBytes.withUnsafeBytes { (buffer: UnsafeRawBufferPointer) -> rac_result_t in
                rac_solution_create_from_proto(buffer.baseAddress, configBytes.count, &raw)
            }

            guard result == RAC_SUCCESS, let raw else {
                throw SDKException(
                    code: .invalidConfiguration,
                    message: SolutionHandle.errorMessage(
                        op: "rac_solution_create_from_proto",
                        rc: result
                    ),
                    category: .internal
                )
            }

            return SolutionHandle(handle: raw)
        }

        /// Convenience overload: serialise a typed `RASolutionConfig`
        /// proto and forward to `run(configBytes:)`. Frontends that
        /// already speak the generated proto types should prefer this
        /// surface; downstream the bytes path is identical.
        public func run(config: RASolutionConfig) async throws -> SolutionHandle {
            let bytes = try config.serializedData()
            return try await run(configBytes: bytes)
        }

        /// YAML sugar — construct a solution from a YAML document. The
        /// loader accepts both `SolutionConfig` shape (top-level oneof
        /// key) and `PipelineSpec` shape (`name`/`operators`/`edges`).
        public func run(yaml: String) async throws -> SolutionHandle {
            try await ensureReady()

            var raw: rac_solution_handle_t?
            let result = yaml.withCString { rac_solution_create_from_yaml($0, &raw) }

            guard result == RAC_SUCCESS, let raw else {
                throw SDKException(
                    code: .invalidConfiguration,
                    message: SolutionHandle.errorMessage(
                        op: "rac_solution_create_from_yaml",
                        rc: result
                    ),
                    category: .internal
                )
            }

            return SolutionHandle(handle: raw)
        }

        private func ensureReady() async throws {
            guard RunAnywhere.isInitialized else {
                throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
            }
            try await RunAnywhere.ensureServicesReady()
        }
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/Solutions/RunAnywhere+Solutions.ts
/**
 * RunAnywhere+Solutions.ts
 *
 * Public API for L5 solutions runtime (T4.7 / T4.8). A "solution" is a
 * prepackaged pipeline config — a typed `SolutionConfig` proto, raw
 * proto bytes, or YAML sugar — that the C++ core compiles into a
 * GraphScheduler DAG and runs through the `rac_solution_*` C ABI.
 *
 * Surface mirrors Swift / Kotlin / Flutter / Web:
 *
 *   const handle = await RunAnywhere.solutions.run({ config })
 *   await handle.start()
 *   await handle.feed('hello')
 *   await handle.closeInput()
 *   await handle.destroy()
 *
 * Reference: sdk/runanywhere-swift/.../Public/Extensions/Solutions/
 */
import { requireNativeModule, isNativeModuleAvailable } from '../../../native';
import { SolutionConfig } from '@runanywhere/proto-ts/solutions';
import { bytesToArrayBuffer } from '../../../services/ProtoBytes';
import { encodeProtoMessage } from '../../../services/ProtoWire';
import { SDKException } from '../../../Foundation/Errors/SDKException';
import { ensureServicesReady } from '../../../Foundation/Initialization/ServicesReadyGuard';
import { requireInitialized } from '../../../Foundation/Initialization/InitializedGuard';

function ensureNative() {
  if (!isNativeModuleAvailable()) {
    throw SDKException.nativeModuleUnavailable();
  }
  return requireNativeModule();
}

/**
 * Lifecycle handle for a started solution.
 *
 * Wraps the C `rac_solution_handle_t` (round-tripped as a JS number).
 * Every verb is async because the native implementation uses Nitro
 * Promises — callers should `await` each call. `destroy()` is
 * idempotent; the handle is inert once destroyed.
 */
export class SolutionHandle {
  private handle: number;
  private alive = true;

  /** @internal */ constructor(handle: number) {
    if (!handle) {
      throw SDKException.invalidInput(
        'Cannot construct SolutionHandle from null native handle'
      );
    }
    this.handle = handle;
  }

  /** True until [destroy] clears the underlying native handle. */
  get isAlive(): boolean {
    return this.alive;
  }

  /** Start the underlying scheduler. Non-blocking. */
  async start(): Promise<void> {
    this.requireAlive();
    const ok = await ensureNative().solutionStart(this.handle);
    // Swift parity: lifecycle-verb failures throw .processingFailed
    // (RunAnywhere+Solutions.swift:108-117).
    if (!ok) throw SDKException.processingFailed('rac_solution_start failed');
  }

  /** Request a graceful shutdown. Non-blocking. */
  async stop(): Promise<void> {
    this.requireAlive();
    const ok = await ensureNative().solutionStop(this.handle);
    if (!ok) throw SDKException.processingFailed('rac_solution_stop failed');
  }

  /** Force-cancel the graph; returns once workers observe cancellation. */
  async cancel(): Promise<void> {
    this.requireAlive();
    const ok = await ensureNative().solutionCancel(this.handle);
    if (!ok) throw SDKException.processingFailed('rac_solution_cancel failed');
  }

  /** Feed one UTF-8 item into the root input edge. */
  async feed(item: string): Promise<void> {
    this.requireAlive();
    const ok = await ensureNative().solutionFeed(this.handle, item);
    if (!ok) throw SDKException.processingFailed('rac_solution_feed failed');
  }

  /** Signal end-of-stream on the root input edge. */
  async closeInput(): Promise<void> {
    this.requireAlive();
    const ok = await ensureNative().solutionCloseInput(this.handle);
    if (!ok) throw SDKException.processingFailed('rac_solution_close_input failed');
  }

  /** Cancel, join, and release native resources. Idempotent. */
  async destroy(): Promise<void> {
    if (!this.alive) return;
    this.alive = false;
    await ensureNative().solutionDestroy(this.handle);
    this.handle = 0;
  }

  private requireAlive(): void {
    if (!this.alive) {
      // Swift parity: using a destroyed handle throws .invalidState
      // (RunAnywhere+Solutions.swift:99-104).
      throw SDKException.invalidState(
        'Solution handle has already been destroyed'
      );
    }
  }
}

/**
 * Discriminated-union argument for [solutions.run].
 *
 * Exactly one input kind is accepted at compile time — the TypeScript
 * compiler rejects call sites that supply more than one of {config,
 * configBytes, yaml}, mirroring the three overloads on Swift/Kotlin.
 */
export type SolutionRunArgs =
  | {
      /** Typed `SolutionConfig` proto — encoded by the SDK before dispatch. */
      config: SolutionConfig;
      configBytes?: never;
      yaml?: never;
    }
  | {
      config?: never;
      /** Raw SolutionConfig / PipelineSpec proto bytes. */
      configBytes: Uint8Array;
      yaml?: never;
    }
  | {
      config?: never;
      configBytes?: never;
      /** YAML sugar (SolutionConfig-shape or PipelineSpec-shape). */
      yaml: string;
    };

/**
 * Construct and return a (created, not yet started) solution. Callers
 * own the returned [SolutionHandle] — invoke `.destroy()` when finished.
 */
async function run(args: SolutionRunArgs): Promise<SolutionHandle> {
  // Swift parity: ensureReady() guards isInitialized
  // (RunAnywhere+Solutions.swift:225-227).
  requireInitialized();
  const native = ensureNative();
  // Swift parity: RunAnywhere+Solutions.swift:228 gates on ensureServicesReady.
  await ensureServicesReady();

  if (args.yaml !== undefined) {
    const h = await native.solutionCreateFromYaml(args.yaml);
    if (!h) {
      // Swift parity: create failures throw .invalidConfiguration
      // (RunAnywhere+Solutions.swift:210-219).
      throw SDKException.invalidConfiguration(
        'rac_solution_create_from_yaml failed'
      );
    }
    return new SolutionHandle(h);
  }

  const configBuffer = args.configBytes
    ? bytesToArrayBuffer(args.configBytes)
    : args.config
      ? encodeProtoMessage(args.config, SolutionConfig)
      : new ArrayBuffer(0);
  if (configBuffer.byteLength === 0) {
    throw SDKException.invalidInput(
      'Solution config bytes are empty — refusing to call rac_solution_create_from_proto'
    );
  }

  const h = await native.solutionCreateFromProto(configBuffer);
  if (!h) {
    // Swift parity: create failures throw .invalidConfiguration
    // (RunAnywhere+Solutions.swift:178-187).
    throw SDKException.invalidConfiguration(
      'rac_solution_create_from_proto failed'
    );
  }
  return new SolutionHandle(h);
}

/**
 * `RunAnywhere.solutions` capability accessor.
 *
 * Stateless — every call to `run(...)` allocates a fresh
 * `rac_solution_handle_t`; callers own the returned [SolutionHandle].
 */
export const solutions = {
  run,
};
